backup/src/commonTest/kotlin/com/wire/backup/encryption/XChaChaPoly1305EncryptedStreamTest.kt

```diff
@@ -0,0 +1,188 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ */
+package com.wire.backup.encryption
+
+import com.ionspin.kotlin.crypto.pwhash.crypto_pwhash_SALTBYTES
+import com.wire.backup.envelope.cryptography.BackupPassphrase
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.test.runTest
+import okio.Buffer
+import okio.ByteString.Companion.encodeUtf8
+import kotlin.random.Random
+import kotlin.test.Ignore
+import kotlin.test.Test
+import kotlin.test.assertContentEquals
+import kotlin.test.assertEquals
+
+class XChaChaPoly1305EncryptedStreamTest {
+
+    private val stream = EncryptedStream.XChaCha20Poly1305
+
+    @Test
+    fun givenEncryptedMessageAndCorrectAuthentication_whenDecrypting_thenShouldReturnOriginalMessage() = runTest {
+        val messageToEncrypt = "Hello Alice!"
+        testCorrectDecryptionOfMessage(messageToEncrypt.encodeToByteArray())
+    }
+
+    private suspend fun testCorrectDecryptionOfMessage(messageToEncrypt: ByteArray) {
+        val originalBuffer = Buffer()
+        originalBuffer.write(messageToEncrypt)
+        val originalBufferData = originalBuffer.peek().readByteArray()
+
+        val encryptedBuffer = Buffer()
+        val passphrase = BackupPassphrase("password")
+        val salt = XChaChaPoly1305AuthenticationData.newSalt()
+        val additionalData = "additionalData".encodeToByteArray().toUByteArray()
+        val authenticationData = XChaChaPoly1305AuthenticationData(passphrase, salt, additionalData)
+        val encryptionHeader = stream.encrypt(
+            originalBuffer,
+            encryptedBuffer,
+            authenticationData
+        )
+
+        val decryptedBuffer = Buffer()
+        val result = stream.decrypt(encryptedBuffer, decryptedBuffer, authenticationData, encryptionHeader)
+        assertEquals(DecryptionResult.Success, result)
+        assertContentEquals(originalBufferData, decryptedBuffer.readByteArray())
+    }
+
+    /**
+     * Disabled for now as it takes more than 5s to run this on CI / JavaScript / Browser and throws a timeout exception
+     * TODO: Refactor and either split into individual tests, or find a way to increase timeout on browser tests. 5s is too short!
+     */
+    @Test
+    @Ignore
+    fun givenEncryptedMessagesOfMultipleSizesAndCorrectAuthentication_whenDecrypting_thenShouldReturnOriginalMessage() = runTest {
```

Comment:
Maybe we should use kotest to support parametrized tests?

Reply:
I didn't investigate all I could about this, to be honest.

It seems to be a bug with browser tests running async (coroutine) code. It spits out a `Timeout of 5000ms exceeded. For async tests and hooks, ensure “done()” is called; if returning a Promise, ensure it resolves.`

I believe it is within JavaScript's execution of `runTest`.

-----------

Maybe a runBlocking would solve it, as it hides the "async" code? Not 100% sure, to be honest. 

-----------

Not sure if it would solve this issue in particular. But in any case, Kotest and parameterised tests could be a nice addition.

Reply:
If we don't have time for this improvement, can we at least create a task for that to not leave ignore annotation? 😄

Reply:
I've split it into smaller tests